src/utils/xhr-utils.js

```diff
@@ -0,0 +1,626 @@
+
+/**
+ * Normalized method name by uppercasing.
+ * @param {string|undefined} method
+ * @return {string}
+ * @private
+ */
+function normalizeMethod_(method) {
+  if (method === undefined) {
+    return 'GET';
+  }
+  method = method.toUpperCase();
+  dev().assert(
+      allowedMethods_.includes(method),
+      'Only one of %s is currently allowed. Got %s',
+      allowedMethods_.join(', '),
+      method
+  );
+  return method;
+}
+
+/**
+ * Verifies if response has the correct headers
+ * @param {!Window} win
+ * @param {!FetchResponse} response
+ * @param {!../utils/xhr-utils.FetchInitDef=} init
+ * @return {!FetchResponse}
+ */
+export function verifyAmpCORSHeaders(win, response, init) {
+  const allowSourceOriginHeader = response.headers.get(
+      ALLOW_SOURCE_ORIGIN_HEADER);
+  if (allowSourceOriginHeader) {
+    const sourceOrigin = getSourceOrigin(win.location.href);
+    // If the `AMP-Access-Control-Allow-Source-Origin` header is returned,
+    // ensure that it's equal to the current source origin.
+    user().assert(allowSourceOriginHeader == sourceOrigin,
+        `Returned ${ALLOW_SOURCE_ORIGIN_HEADER} is not` +
+          ` equal to the current: ${allowSourceOriginHeader}` +
+          ` vs ${sourceOrigin}`);
+  } else if (init.requireAmpResponseSourceOrigin) {
+    // If the `AMP-Access-Control-Allow-Source-Origin` header is not
+    // returned but required, return error.
+    user().assert(false, 'Response must contain the' +
+        ` ${ALLOW_SOURCE_ORIGIN_HEADER} header`);
+  }
+  return response;
+}
+
+// TODO(prateekbh): move everything below this line into the polyfill
+/**
+ * A record version of `XMLHttpRequest` that has all the necessary properties
+ * and methods of `XMLHttpRequest` to construct a `FetchResponse` from a
+ * serialized response returned by the viewer.
+ * @typedef {{
+ *   status: number,
+ *   statusText: string,
+ *   responseText: string,
+ *   responseXML: ?Document,
+ *   getResponseHeader: function(this:XMLHttpRequestDef, string): string,
+ * }}
+ */
+let XMLHttpRequestDef;
+
+/** @private @enum {number} Allowed fetch responses. */
+const allowedFetchTypes_ = {
+  document: 1,
+  text: 2,
+};
+
+/**
+ * A minimal polyfill of Fetch API. It only polyfills what we currently use.
+ *
+ * See https://developer.mozilla.org/en-US/docs/Web/API/GlobalFetch/fetch
+ *
+ * Notice that the "fetch" method itself is not exported as that would require
+ * us to immediately support a much wide API.
+ *
+ * @param {string} input
+ * @param {!FetchInitDef} init
+ * @return {!Promise<!FetchResponse>}
+ * @private Visible for testing
+ */
+export function fetchPolyfill(input, init) {
+  return new Promise(function(resolve, reject) {
+    const xhr = createXhrRequest(init.method || 'GET', input);
+
+    if (init.credentials == 'include') {
+      xhr.withCredentials = true;
+    }
+
+    if (init.responseType in allowedFetchTypes_) {
+      xhr.responseType = init.responseType;
+    }
+
+    if (init.headers) {
```

Comment:
Nit: This should be [guaranteed](https://github.com/ampproject/amphtml/pull/17206/files#diff-6870dab961227bda821a287600fb7158R281).